{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "50abf7b9-5395-46e4-a457-cf430a9b674e",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-02T07:00:31.440882Z",
     "iopub.status.busy": "2024-12-02T07:00:31.438040Z",
     "iopub.status.idle": "2024-12-02T07:00:36.729242Z",
     "shell.execute_reply": "2024-12-02T07:00:36.724778Z"
    },
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "library(\"limma\")\n",
    "library(\"dplyr\")\n",
    "library(\"tidyverse\")\n",
    "library(\"caret\")\n",
    "library(\"switchBox\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bc657e02-4938-4cda-9d5d-848472334341",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "# 1、Data Processing"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c07acc39-f15a-4ae6-a45a-fd9e984d4f24",
   "metadata": {},
   "source": [
    "**Steps:**\n",
    "1. Data preprocessing\n",
    "2. Split training, testing, and external independent validation sets"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "63eb23e5-cd75-495f-8c3d-51c405e47d00",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### 1.1 Data preprocessing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "46e73288-ff90-42b5-b1f5-23ec69318e40",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-11-28T11:05:53.414247Z",
     "iopub.status.busy": "2024-11-28T11:05:53.411792Z",
     "iopub.status.idle": "2024-11-28T11:05:53.440336Z",
     "shell.execute_reply": "2024-11-28T11:05:53.433169Z"
    }
   },
   "outputs": [],
   "source": [
    "# Define chip data processing function\n",
    "array_data_process = function(data_path, expr, probeid, other_columns, gpl_file_path, genetosymbol_file_path, colnames_file_path, output_file) {\n",
    "  # Read and preprocess expression data\n",
    "  data = read.ilmn(data_path, expr = expr, probeid = probeid, other.columns = other_columns)\n",
    "  data = neqc(data, detection.p = other_columns)\n",
    "  data = data$E\n",
    "  data = cbind(rownames(data), data)\n",
    "  colnames(data)[1] = \"ID\"\n",
    "\n",
    "  # Read gpl and genetosymbol file and filter\n",
    "  GPL = read.table(gpl_file_path, header = TRUE, sep = \"\\t\", fill = TRUE, comment.char = \"#\", quote = \"\\\"\")\n",
    "  GPL = GPL %>% select(ID, Source, Entrez_Gene_ID) %>% filter(Source == \"RefSeq\") %>% select(ID, Entrez_Gene_ID)\n",
    "  genetosymbol = read.table(genetosymbol_file_path, header = FALSE, sep = \"\\t\", fill = TRUE, quote = \"\\\"\")\n",
    "  genetosymbol = genetosymbol[,c(2,3)]\n",
    "  colnames(genetosymbol) = c(\"Entrez_Gene_ID\", \"Symbol\")\n",
    "    \n",
    "  # Remove duplicate gene symbols\n",
    "  data = merge(GPL, data, by = \"ID\")\n",
    "  data = merge(genetosymbol, data, by = \"Entrez_Gene_ID\")\n",
    "  data = data[, -c(1, 3)]\n",
    "  data[,-1] = apply(data[,-1], 2, as.numeric)\n",
    "  index = order(rowMeans(data[,-1]), decreasing = TRUE)\n",
    "  data = data[index, ]\n",
    "  keep = !duplicated(data$Symbol)\n",
    "  data = data[keep, ]\n",
    "  rownames(data) = data[, 1]\n",
    "  data = as.matrix(data[, -1])\n",
    "\n",
    "  # Rename column names\n",
    "  Colnames = read.table(colnames_file_path, header = FALSE, sep = \"\\t\")\n",
    "  match_result = match(Colnames[, 1], colnames(data))\n",
    "  data = data[, match_result]\n",
    "  colnames(data) = Colnames[, 2]\n",
    "\n",
    "  # Save processed data\n",
    "  write.table(data, output_file, sep = \"\\t\")\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "f3e2fea3-7014-4f0d-ad69-bc09a83e60cf",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-11-28T11:06:14.477898Z",
     "iopub.status.busy": "2024-11-28T11:06:14.475815Z",
     "iopub.status.idle": "2024-11-28T11:06:47.424823Z",
     "shell.execute_reply": "2024-11-28T11:06:47.419572Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reading file ./data/01_Raw_data/GSE19444/GSE19444_non-normalized.txt ... ...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Note: inferring mean and variance of negative control probe intensities from the detection p-values.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Function call\n",
    "data_path = \"./data/01_Raw_data/GSE19444/GSE19444_non-normalized.txt\"\n",
    "expr = \"VALUE_\"\n",
    "probeid = \"ID_REF\"\n",
    "other_columns = \"Detection PVal\"\n",
    "gpl_file_path = \"./data/01_Raw_data/GSE19439/GPL6947-13512.txt\"\n",
    "genetosymbol_file_path = \"./data/01_Raw_data/GSE19439/genetosymbol.txt\"\n",
    "colnames_file_path = \"./data/01_Raw_data/GSE19444/Colnames.txt\"\n",
    "output_file = \"./data/02_Preliminary_data/GSE19444/data.txt\"\n",
    "array_data_process(data_path, expr, probeid, other_columns, gpl_file_path, genetosymbol_file_path, colnames_file_path, output_file)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7ab1ff63-274f-4da1-a1e9-44e53fee8a71",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### 1.2 Split training, testing, and external independent validation sets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "616dbd43-ef3c-4ca2-8954-a1293934500c",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-11-28T11:12:36.303816Z",
     "iopub.status.busy": "2024-11-28T11:12:36.300983Z",
     "iopub.status.idle": "2024-11-28T11:12:48.716514Z",
     "shell.execute_reply": "2024-11-28T11:12:48.714002Z"
    },
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Data loaded with name: GSE19439\"\n",
      "[1] \"Data loaded with name: GSE19442\"\n",
      "[1] \"Data loaded with name: GSE19444\"\n",
      "[1] \"Data loaded with name: GSE83456\"\n"
     ]
    }
   ],
   "source": [
    "# Merge data\n",
    "base_path = \"./data/02_Preliminary_data/\"\n",
    "folder_list = list.files(base_path, full.names = TRUE)\n",
    "data = list()\n",
    "for (folder in folder_list) {\n",
    "  folder_name = basename(folder)\n",
    "  file_path = file.path(folder, \"data.txt\")\n",
    "  if (file.exists(file_path)) {\n",
    "    data_tmp = read.table(file_path, header = TRUE, sep = \"\\t\", check.names = FALSE)\n",
    "    data_tmp = cbind(rownames(data_tmp), data_tmp)\n",
    "    colnames(data_tmp)[1] = \"Symbol\"\n",
    "    data[[folder_name]] = data_tmp\n",
    "    print(paste(\"Data loaded with name:\", folder_name))\n",
    "  } else {\n",
    "    cat(\"File not found:\", folder_name, \"\\n\")\n",
    "  }\n",
    "}    \n",
    "all_data = Reduce(function(x, y) merge(x, y, by = \"Symbol\"), data)\n",
    "rownames(all_data) = all_data[,1]\n",
    "all_data = all_data[,-1]\n",
    "all_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "584a694a-101c-4e52-8033-a65893dc1763",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-11-28T11:15:14.413927Z",
     "iopub.status.busy": "2024-11-28T11:15:14.412022Z",
     "iopub.status.idle": "2024-11-28T11:15:28.508090Z",
     "shell.execute_reply": "2024-11-28T11:15:28.503962Z"
    }
   },
   "outputs": [],
   "source": [
    "# Split training and testing sets\n",
    "selected_cols = grep(\"^(01|02|03)\", colnames(all_data))  \n",
    "Train_Test_data = all_data[, selected_cols]\n",
    "Train_Test_data = t(Train_Test_data)\n",
    "label = ifelse(grepl(\"_PTB_\", rownames(Train_Test_data)), 1, 0)\n",
    "Train_Test_data = as.data.frame(cbind(Train_Test_data, label = label))\n",
    "set.seed(123)  \n",
    "split_index = createDataPartition(Train_Test_data$label, p = 0.7, list = FALSE)\n",
    "train_data = Train_Test_data[split_index, ]\n",
    "test_data = Train_Test_data[-split_index, ]\n",
    "write.csv(train_data, \"./data/03_Final_data/01_train_data.csv\")\n",
    "write.csv(test_data, \"./data/03_Final_data/02_test_data.csv\")\n",
    "\n",
    "# Split external independent validation set\n",
    "selected_cols = grep(\"^(04)\", colnames(all_data))  \n",
    "valid_data = all_data[, selected_cols]\n",
    "valid_data = t(valid_data)\n",
    "label = ifelse(grepl(\"_PTB_\", rownames(valid_data)), 1, 0)\n",
    "valid_data = as.data.frame(cbind(valid_data, label = label))\n",
    "write.csv(valid_data, \"./data/03_Final_data/03_valid_data.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "093638ff-0364-4ed8-bfe4-e3d9d435ee04",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "# 2、Gene pair methods based on gene expression values"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "20e1485d-a033-4e38-9bcc-677b2856757a",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### 2.1 GERs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0976d13b-60cf-4415-a05e-05628cd2da58",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "#### 2.1.1 Train"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "28e27306-29c3-4079-b527-f6458895011d",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-11-28T11:16:42.057627Z",
     "iopub.status.busy": "2024-11-28T11:16:42.054498Z",
     "iopub.status.idle": "2024-11-28T11:16:42.077000Z",
     "shell.execute_reply": "2024-11-28T11:16:42.075606Z"
    }
   },
   "outputs": [],
   "source": [
    "# Define differential analysis function\n",
    "performDEGAnalysis = function(filepath, prefixes, classifyString, adjPValThreshold, logFCThreshold, Samples, limma_results_filepath, DEG_results_filepath) {\n",
    "  # Read and process data\n",
    "  DEGAnalysis_data = read.csv(filepath, check.names = FALSE, row.names = 1)\n",
    "  DEGAnalysis_data = DEGAnalysis_data[, -ncol(DEGAnalysis_data)]\n",
    "  DEGAnalysis_data = t(DEGAnalysis_data)\n",
    "  DEGAnalysis_data = log2(DEGAnalysis_data + 1)\n",
    "    \n",
    "  # Initialize two empty data frames to store all results\n",
    "  limma_results = data.frame()\n",
    "  DEG_results = data.frame(Data = character(), Gene = character(), adjust_P_val = numeric(), logFC = numeric(), stringsAsFactors = FALSE)\n",
    "  \n",
    "  i = 1\n",
    "  for(prefix in prefixes) {\n",
    "    print(prefix)\n",
    "\n",
    "    # Extract data\n",
    "    col_names = grep(paste0(\"^\", prefix), colnames(DEGAnalysis_data), value = TRUE)\n",
    "    data = DEGAnalysis_data[, col_names]\n",
    "    data = data[which(rowSums(data) != 0), ]\n",
    "\n",
    "    # Change order\n",
    "    pos_columns = grep(classifyString, colnames(data), value = TRUE)\n",
    "    neg_columns = setdiff(colnames(data), pos_columns)\n",
    "    new_column_order = c(pos_columns, neg_columns)\n",
    "    data = data[, new_column_order]\n",
    "      \n",
    "    # Create group list\n",
    "    total_columns = dim(data)[2]\n",
    "    count_contains = sum(grepl(classifyString, colnames(data)))\n",
    "    count_not_contains = total_columns - count_contains\n",
    "    group_list = c(rep('Pos', count_contains), rep('Neg', count_not_contains))\n",
    "    \n",
    "    # Create design matrix\n",
    "    design = model.matrix(~0 + factor(group_list))\n",
    "    colnames(design) = levels(factor(group_list))\n",
    "    rownames(design) = colnames(data)\n",
    "    \n",
    "    # Perform differential expression analysis\n",
    "    fit = lmFit(data, design)\n",
    "    contrast.matrix = makeContrasts(Pos-Neg, levels = design)\n",
    "    fit2 = contrasts.fit(fit, contrast.matrix)\n",
    "    fit2 = eBayes(fit2)\n",
    "    \n",
    "    # Extract differentially expressed genes\n",
    "    DEG = topTable(fit2, n = Inf)\n",
    "    DEG = na.omit(DEG)\n",
    "    DEG_filter = DEG[DEG$adj.P.Val < adjPValThreshold & abs(DEG$logFC) > logFCThreshold, ]\n",
    "    \n",
    "    # Construct current analysis results data frame and append \n",
    "    limma_results = rbind(limma_results, DEG)\n",
    "    tempResults = data.frame(Sample = rep(Samples[i], nrow(DEG_filter)), Gene = row.names(DEG_filter), adjust_P_val = DEG_filter$adj.P.Val, logFC = DEG_filter$logFC, stringsAsFactors = FALSE)\n",
    "    DEG_results = rbind(DEG_results, tempResults)\n",
    "    i = i + 1\n",
    "  }\n",
    "    \n",
    "  # Output results\n",
    "  write.csv(limma_results, limma_results_filepath, row.names = TRUE)\n",
    "  write.csv(DEG_results, DEG_results_filepath, row.names = FALSE)\n",
    "  return(list(limma_results = limma_results, DEG_results = DEG_results))\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "8dae6a5d-e3ff-4b3a-93d1-fa035d092d70",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-11-28T11:25:12.694654Z",
     "iopub.status.busy": "2024-11-28T11:25:12.689580Z",
     "iopub.status.idle": "2024-11-28T11:25:29.229825Z",
     "shell.execute_reply": "2024-11-28T11:25:29.226821Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"0\"\n"
     ]
    }
   ],
   "source": [
    "# Call function to perform differential analysis\n",
    "filepath = \"./data/03_Final_data/01_train_data.csv\"\n",
    "prefixes = c(\"0\")\n",
    "classifyString = \"_PTB_\"\n",
    "adjPValThreshold = 0.0000000015\n",
    "logFCThreshold = 0\n",
    "Samples = c(\"train_data\")\n",
    "limma_results_filepath = \"./result/01_GERs/01_limma_results.csv\"\n",
    "DEG_results_filepath = \"./result/01_GERs/02_limma_DEG_results.csv\"\n",
    "Results = performDEGAnalysis(filepath, prefixes, classifyString, adjPValThreshold, logFCThreshold, Samples, limma_results_filepath, DEG_results_filepath)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "809e2ffc-5db7-4a39-b49d-7b7f2b3b695d",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-11-28T11:26:11.465014Z",
     "iopub.status.busy": "2024-11-28T11:26:11.463987Z",
     "iopub.status.idle": "2024-11-28T11:26:11.482366Z",
     "shell.execute_reply": "2024-11-28T11:26:11.479620Z"
    }
   },
   "outputs": [],
   "source": [
    "# Save ratio pairs\n",
    "upregulated = Results$DEG_results$Gene[Results$DEG_results$logFC > 0]\n",
    "downregulated = Results$DEG_results$Gene[Results$DEG_results$logFC < 0]\n",
    "result = expand.grid(Upregulated = upregulated, Downregulated = downregulated)\n",
    "write.csv(result, \"./result/01_GERs/03_ratio_pair.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c083f7ac-110d-4441-90fd-4b42b343e3c7",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "#### 2.1.2 Test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "b86e0a85-3bc7-4e0f-b0ec-f60297fc9c7a",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-11-28T13:03:58.699594Z",
     "iopub.status.busy": "2024-11-28T13:03:58.697826Z",
     "iopub.status.idle": "2024-11-28T13:03:58.722547Z",
     "shell.execute_reply": "2024-11-28T13:03:58.717012Z"
    },
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Define accuracy calculation function\n",
    "calculate_accuracy = function(df, ratio_pair) {\n",
    "  # Extract labels\n",
    "  labels = df[, ncol(df)]\n",
    "  df = df[, -ncol(df)]\n",
    "  \n",
    "  # Get number of samples and combinations\n",
    "  num_samples = nrow(df)\n",
    "  num_pairs = nrow(ratio_pair)\n",
    "  \n",
    "  # Store prediction results\n",
    "  predictions = numeric(num_samples)\n",
    "  \n",
    "  # Iterate through samples\n",
    "  for (i in 1:num_samples) {\n",
    "    sample_data = df[i, ]\n",
    "    count_positive = 0\n",
    "    \n",
    "    for (j in 1:num_pairs) {\n",
    "      gene1 = ratio_pair[j, 1]\n",
    "      gene2 = ratio_pair[j, 2]\n",
    "\n",
    "      if (sample_data[gene1] > sample_data[gene2]) {\n",
    "        count_positive = count_positive + 1\n",
    "      }\n",
    "    }\n",
    "    \n",
    "    # Determine if positive class based on condition\n",
    "    if (count_positive > num_pairs / 2) {\n",
    "      predictions[i] = 1  \n",
    "    } else {\n",
    "      predictions[i] = 0  \n",
    "    }\n",
    "  }\n",
    "  \n",
    "  # Calculate accuracy\n",
    "  accuracy = sum(predictions == labels) / num_samples\n",
    "  \n",
    "  # Output accuracy\n",
    "  return(accuracy)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "20a2008d-e9fb-4372-93c3-05fb69cb629b",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-11-28T13:05:22.021119Z",
     "iopub.status.busy": "2024-11-28T13:05:22.017950Z",
     "iopub.status.idle": "2024-11-28T13:06:13.407886Z",
     "shell.execute_reply": "2024-11-28T13:06:13.405124Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "0.844660194174757"
      ],
      "text/latex": [
       "0.844660194174757"
      ],
      "text/markdown": [
       "0.844660194174757"
      ],
      "text/plain": [
       "[1] 0.8446602"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Call function to calculate accuracy\n",
    "df = read.csv(\"./data/03_Final_data/01_train_data.csv\", check.names = FALSE, row.names = 1)\n",
    "ratio_pair = read.csv(\"./result/01_GERs/03_ratio_pair.csv\", check.names = FALSE, row.names = 1)\n",
    "accuracy = calculate_accuracy(df, ratio_pair)\n",
    "accuracy"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6a09685a-88f0-477c-a2be-eadc9cace11e",
   "metadata": {},
   "source": [
    "**train:0.845 ; test:0.818 ; valid:0.849**"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5cf5ff7e-95d5-4409-aaaa-75b713756500",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "# 3、Gene pair methods based on gene ranking relationships"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1149a229-d916-42d6-bfef-eb09da388677",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### 3.1 k-TSP"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "07d35a50-c2b7-45ee-8124-b738e4b87621",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "#### 3.1.1 Train"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "c8d5aa7f-0048-46e1-be24-222676c2b3a4",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-01T08:11:05.698852Z",
     "iopub.status.busy": "2024-12-01T08:11:05.695919Z",
     "iopub.status.idle": "2024-12-01T08:11:24.997785Z",
     "shell.execute_reply": "2024-12-01T08:11:24.993386Z"
    },
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Import data\n",
    "train_data = read.csv(\"./data/03_Final_data/01_train_data.csv\", check.names = FALSE, row.names = 1)\n",
    "train_labels = as.factor(train_data[, ncol(train_data)])\n",
    "train_data = train_data[, -col(train_data)]\n",
    "train_data = t(train_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "10e55144-6986-4062-8217-c6367cc7f855",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-01T08:11:25.006416Z",
     "iopub.status.busy": "2024-12-01T08:11:25.003190Z",
     "iopub.status.idle": "2024-12-01T08:11:27.399418Z",
     "shell.execute_reply": "2024-12-01T08:11:27.395939Z"
    },
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<dl>\n",
       "\t<dt>$name</dt>\n",
       "\t\t<dd>'5TSPS'</dd>\n",
       "\t<dt>$TSPs</dt>\n",
       "\t\t<dd><table class=\"dataframe\">\n",
       "<caption>A matrix: 5 × 2 of type chr</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>gene1</th><th scope=col>gene2</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>ANKRD22,KLF12</th><td>ANKRD22</td><td>KLF12 </td></tr>\n",
       "\t<tr><th scope=row>PSTPIP2,SLC4A7</th><td>PSTPIP2</td><td>SLC4A7</td></tr>\n",
       "\t<tr><th scope=row>GBP4,WDR75</th><td>GBP4   </td><td>WDR75 </td></tr>\n",
       "\t<tr><th scope=row>FCGR1BP,SLAMF6</th><td>FCGR1BP</td><td>SLAMF6</td></tr>\n",
       "\t<tr><th scope=row>GBP6,ADAM7</th><td>GBP6   </td><td>ADAM7 </td></tr>\n",
       "</tbody>\n",
       "</table>\n",
       "</dd>\n",
       "\t<dt>$score</dt>\n",
       "\t\t<dd><style>\n",
       ".dl-inline {width: auto; margin:0; padding: 0}\n",
       ".dl-inline>dt, .dl-inline>dd {float: none; width: auto; display: inline-block}\n",
       ".dl-inline>dt::after {content: \":\\0020\"; padding-right: .5ex}\n",
       ".dl-inline>dt:not(:first-of-type) {padding-left: .5ex}\n",
       "</style><dl class=dl-inline><dt>ANKRD22,KLF12</dt><dd>0.868287355924979</dd><dt>PSTPIP2,SLC4A7</dt><dd>0.853794306479113</dd><dt>GBP4,WDR75</dt><dd>0.810316075873828</dd><dt>FCGR1BP,SLAMF6</dt><dd>0.809464251918159</dd><dt>GBP6,ADAM7</dt><dd>0.809463319266837</dd></dl>\n",
       "</dd>\n",
       "\t<dt>$tieVote</dt>\n",
       "\t\t<dd><style>\n",
       ".dl-inline {width: auto; margin:0; padding: 0}\n",
       ".dl-inline>dt, .dl-inline>dd {float: none; width: auto; display: inline-block}\n",
       ".dl-inline>dt::after {content: \":\\0020\"; padding-right: .5ex}\n",
       ".dl-inline>dt:not(:first-of-type) {padding-left: .5ex}\n",
       "</style><dl class=dl-inline><dt>ANKRD22,KLF12</dt><dd>both</dd><dt>PSTPIP2,SLC4A7</dt><dd>both</dd><dt>GBP4,WDR75</dt><dd>both</dd><dt>FCGR1BP,SLAMF6</dt><dd>both</dd><dt>GBP6,ADAM7</dt><dd>both</dd></dl>\n",
       "\n",
       "<details>\n",
       "\t<summary style=display:list-item;cursor:pointer>\n",
       "\t\t<strong>Levels</strong>:\n",
       "\t</summary>\n",
       "\t<style>\n",
       "\t.list-inline {list-style: none; margin:0; padding: 0}\n",
       "\t.list-inline>li {display: inline-block}\n",
       "\t.list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "\t</style>\n",
       "\t<ol class=list-inline><li>'both'</li><li>'0'</li><li>'1'</li></ol>\n",
       "</details></dd>\n",
       "\t<dt>$labels</dt>\n",
       "\t\t<dd><style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'0'</li><li>'1'</li></ol>\n",
       "</dd>\n",
       "</dl>\n"
      ],
      "text/latex": [
       "\\begin{description}\n",
       "\\item[\\$name] '5TSPS'\n",
       "\\item[\\$TSPs] A matrix: 5 × 2 of type chr\n",
       "\\begin{tabular}{r|ll}\n",
       "  & gene1 & gene2\\\\\n",
       "\\hline\n",
       "\tANKRD22,KLF12 & ANKRD22 & KLF12 \\\\\n",
       "\tPSTPIP2,SLC4A7 & PSTPIP2 & SLC4A7\\\\\n",
       "\tGBP4,WDR75 & GBP4    & WDR75 \\\\\n",
       "\tFCGR1BP,SLAMF6 & FCGR1BP & SLAMF6\\\\\n",
       "\tGBP6,ADAM7 & GBP6    & ADAM7 \\\\\n",
       "\\end{tabular}\n",
       "\n",
       "\\item[\\$score] \\begin{description*}\n",
       "\\item[ANKRD22,KLF12] 0.868287355924979\n",
       "\\item[PSTPIP2,SLC4A7] 0.853794306479113\n",
       "\\item[GBP4,WDR75] 0.810316075873828\n",
       "\\item[FCGR1BP,SLAMF6] 0.809464251918159\n",
       "\\item[GBP6,ADAM7] 0.809463319266837\n",
       "\\end{description*}\n",
       "\n",
       "\\item[\\$tieVote] \\begin{description*}\n",
       "\\item[ANKRD22,KLF12] both\n",
       "\\item[PSTPIP2,SLC4A7] both\n",
       "\\item[GBP4,WDR75] both\n",
       "\\item[FCGR1BP,SLAMF6] both\n",
       "\\item[GBP6,ADAM7] both\n",
       "\\end{description*}\n",
       "\n",
       "\\emph{Levels}: \\begin{enumerate*}\n",
       "\\item 'both'\n",
       "\\item '0'\n",
       "\\item '1'\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\item[\\$labels] \\begin{enumerate*}\n",
       "\\item '0'\n",
       "\\item '1'\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\end{description}\n"
      ],
      "text/markdown": [
       "$name\n",
       ":   '5TSPS'\n",
       "$TSPs\n",
       ":   \n",
       "A matrix: 5 × 2 of type chr\n",
       "\n",
       "| <!--/--> | gene1 | gene2 |\n",
       "|---|---|---|\n",
       "| ANKRD22,KLF12 | ANKRD22 | KLF12  |\n",
       "| PSTPIP2,SLC4A7 | PSTPIP2 | SLC4A7 |\n",
       "| GBP4,WDR75 | GBP4    | WDR75  |\n",
       "| FCGR1BP,SLAMF6 | FCGR1BP | SLAMF6 |\n",
       "| GBP6,ADAM7 | GBP6    | ADAM7  |\n",
       "\n",
       "\n",
       "$score\n",
       ":   ANKRD22,KLF12\n",
       ":   0.868287355924979PSTPIP2,SLC4A7\n",
       ":   0.853794306479113GBP4,WDR75\n",
       ":   0.810316075873828FCGR1BP,SLAMF6\n",
       ":   0.809464251918159GBP6,ADAM7\n",
       ":   0.809463319266837\n",
       "\n",
       "\n",
       "$tieVote\n",
       ":   ANKRD22,KLF12\n",
       ":   bothPSTPIP2,SLC4A7\n",
       ":   bothGBP4,WDR75\n",
       ":   bothFCGR1BP,SLAMF6\n",
       ":   bothGBP6,ADAM7\n",
       ":   both\n",
       "\n",
       "\n",
       "**Levels**: 1. 'both'\n",
       "2. '0'\n",
       "3. '1'\n",
       "\n",
       "\n",
       "\n",
       "$labels\n",
       ":   1. '0'\n",
       "2. '1'\n",
       "\n",
       "\n",
       "\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "$name\n",
       "[1] \"5TSPS\"\n",
       "\n",
       "$TSPs\n",
       "               gene1     gene2   \n",
       "ANKRD22,KLF12  \"ANKRD22\" \"KLF12\" \n",
       "PSTPIP2,SLC4A7 \"PSTPIP2\" \"SLC4A7\"\n",
       "GBP4,WDR75     \"GBP4\"    \"WDR75\" \n",
       "FCGR1BP,SLAMF6 \"FCGR1BP\" \"SLAMF6\"\n",
       "GBP6,ADAM7     \"GBP6\"    \"ADAM7\" \n",
       "\n",
       "$score\n",
       " ANKRD22,KLF12 PSTPIP2,SLC4A7     GBP4,WDR75 FCGR1BP,SLAMF6     GBP6,ADAM7 \n",
       "     0.8682874      0.8537943      0.8103161      0.8094643      0.8094633 \n",
       "\n",
       "$tieVote\n",
       " ANKRD22,KLF12 PSTPIP2,SLC4A7     GBP4,WDR75 FCGR1BP,SLAMF6     GBP6,ADAM7 \n",
       "          both           both           both           both           both \n",
       "Levels: both 0 1\n",
       "\n",
       "$labels\n",
       "[1] \"0\" \"1\"\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Get feature pairs\n",
    "classifier = SWAP.Train.KTSP(train_data, train_labels, krange=c(1, 3, 5))\n",
    "write.csv(classifier$TSPs, \"./result/03_k-TSP/01_kTSP.csv\", row.names = FALSE)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cd3cc52f-a6e0-4753-8de8-60888ef2f77a",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "#### 3.1.2 Test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "c29d52c8-6ec1-405c-bf71-a11b7c0afd93",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-01T08:16:40.186196Z",
     "iopub.status.busy": "2024-12-01T08:16:40.182668Z",
     "iopub.status.idle": "2024-12-01T08:16:57.080332Z",
     "shell.execute_reply": "2024-12-01T08:16:57.076988Z"
    }
   },
   "outputs": [],
   "source": [
    "# Import data\n",
    "test_data = read.csv(\"./data/03_Final_data/02_test_data.csv\", check.names = FALSE, row.names = 1)\n",
    "test_labels = as.factor(test_data[, ncol(test_data)])\n",
    "test_data = test_data[, -ncol(test_data)]\n",
    "test_data = t(test_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "e9ad6b25-f144-4cbf-b374-ab5c9c64188c",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-01T08:17:22.523918Z",
     "iopub.status.busy": "2024-12-01T08:17:22.518158Z",
     "iopub.status.idle": "2024-12-01T08:17:22.575245Z",
     "shell.execute_reply": "2024-12-01T08:17:22.570077Z"
    },
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "0.909090909090909"
      ],
      "text/latex": [
       "0.909090909090909"
      ],
      "text/markdown": [
       "0.909090909090909"
      ],
      "text/plain": [
       "[1] 0.9090909"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Test\n",
    "testPrediction = SWAP.KTSP.Classify(test_data, classifier)\n",
    "accuracy = sum(as.vector(testPrediction) == test_labels) / length(test_labels)\n",
    "accuracy"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e0f839bf-31be-4e5a-9e80-9450f1784635",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-11-29T06:55:13.404648Z",
     "iopub.status.busy": "2024-11-29T06:55:13.402604Z",
     "iopub.status.idle": "2024-11-29T06:55:13.423560Z",
     "shell.execute_reply": "2024-11-29T06:55:13.417100Z"
    }
   },
   "source": [
    "**train:0.942 ; test:0.909 ; valid:0.953**"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "57771385-2b84-4c1c-937f-b032492b1b4f",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### 3.2 TSPG"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7013456c-63a3-48b2-a54e-501169c4d912",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "#### 3.2.1 Train"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "d28b3f0b-8b71-4160-ac6d-bc5c5eedf208",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-02T07:00:36.777426Z",
     "iopub.status.busy": "2024-12-02T07:00:36.774545Z",
     "iopub.status.idle": "2024-12-02T07:00:36.804973Z",
     "shell.execute_reply": "2024-12-02T07:00:36.801109Z"
    }
   },
   "outputs": [],
   "source": [
    "# Define function\n",
    "extract_genes = function(train_data, train_labels, percentage, features_num, iterations_num, freq_threshold, output_file) {\n",
    "  # Get number of samples        \n",
    "  n_samples = ncol(train_data)\n",
    "  n_train = floor(percentage * n_samples)\n",
    "  \n",
    "  # Initialize A and B lists\n",
    "  A_list = list()\n",
    "  B_list = list()\n",
    "  \n",
    "  # Repeat n times\n",
    "  for (i in 1:iterations_num) {\n",
    "    # Randomly sample training data\n",
    "    print(i)\n",
    "    flush.console()\n",
    "    set.seed(i)           \n",
    "    indices = sample(1:n_samples, n_train)\n",
    "    data = train_data[, indices]\n",
    "    labels = train_labels[indices]\n",
    "    \n",
    "    # Get feature gene pairs\n",
    "    classifier = SWAP.Train.KTSP(data, labels, krange = features_num)\n",
    "    tsp_matrix = classifier$TSPs \n",
    "    A_list[[i]] = tsp_matrix[, 1] \n",
    "    B_list[[i]] = tsp_matrix[, 2] \n",
    "  }\n",
    "  \n",
    "  # Merge A and B lists\n",
    "  A_genes = unlist(A_list)\n",
    "  B_genes = unlist(B_list)\n",
    "  \n",
    "  # Calculate frequencies\n",
    "  A_freq = table(A_genes) / iterations_num\n",
    "  B_freq = table(B_genes) / iterations_num\n",
    "  \n",
    "  # Keep genes with frequencies above the threshold\n",
    "  A_selected = names(A_freq[A_freq > freq_threshold])\n",
    "  B_selected = names(B_freq[B_freq > freq_threshold])\n",
    "  \n",
    "  # Create result matrix\n",
    "  result_matrix = cbind(A = A_selected, B = B_selected)\n",
    "  \n",
    "  # Save matrix as file\n",
    "  write.csv(result_matrix, file = output_file)\n",
    "  \n",
    "  return(result_matrix)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "4634172f-a98d-47ec-8c33-4ed82208b24a",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-02T07:00:36.813113Z",
     "iopub.status.busy": "2024-12-02T07:00:36.810217Z",
     "iopub.status.idle": "2024-12-02T07:47:24.223079Z",
     "shell.execute_reply": "2024-12-02T07:47:24.219079Z"
    },
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 1\n",
      "[1] 2\n",
      "[1] 3\n",
      "[1] 4\n",
      "[1] 5\n",
      "[1] 6\n",
      "[1] 7\n",
      "[1] 8\n",
      "[1] 9\n",
      "[1] 10\n",
      "[1] 11\n",
      "[1] 12\n",
      "[1] 13\n",
      "[1] 14\n",
      "[1] 15\n",
      "[1] 16\n",
      "[1] 17\n",
      "[1] 18\n",
      "[1] 19\n",
      "[1] 20\n",
      "[1] 21\n",
      "[1] 22\n",
      "[1] 23\n",
      "[1] 24\n",
      "[1] 25\n",
      "[1] 26\n",
      "[1] 27\n",
      "[1] 28\n",
      "[1] 29\n",
      "[1] 30\n",
      "[1] 31\n",
      "[1] 32\n",
      "[1] 33\n",
      "[1] 34\n",
      "[1] 35\n",
      "[1] 36\n",
      "[1] 37\n",
      "[1] 38\n",
      "[1] 39\n",
      "[1] 40\n",
      "[1] 41\n",
      "[1] 42\n",
      "[1] 43\n",
      "[1] 44\n",
      "[1] 45\n",
      "[1] 46\n",
      "[1] 47\n",
      "[1] 48\n",
      "[1] 49\n",
      "[1] 50\n",
      "[1] 51\n",
      "[1] 52\n",
      "[1] 53\n",
      "[1] 54\n",
      "[1] 55\n",
      "[1] 56\n",
      "[1] 57\n",
      "[1] 58\n",
      "[1] 59\n",
      "[1] 60\n",
      "[1] 61\n",
      "[1] 62\n",
      "[1] 63\n",
      "[1] 64\n",
      "[1] 65\n",
      "[1] 66\n",
      "[1] 67\n",
      "[1] 68\n",
      "[1] 69\n",
      "[1] 70\n",
      "[1] 71\n",
      "[1] 72\n",
      "[1] 73\n",
      "[1] 74\n",
      "[1] 75\n",
      "[1] 76\n",
      "[1] 77\n",
      "[1] 78\n",
      "[1] 79\n",
      "[1] 80\n",
      "[1] 81\n",
      "[1] 82\n",
      "[1] 83\n",
      "[1] 84\n",
      "[1] 85\n",
      "[1] 86\n",
      "[1] 87\n",
      "[1] 88\n",
      "[1] 89\n",
      "[1] 90\n",
      "[1] 91\n",
      "[1] 92\n",
      "[1] 93\n",
      "[1] 94\n",
      "[1] 95\n",
      "[1] 96\n",
      "[1] 97\n",
      "[1] 98\n",
      "[1] 99\n",
      "[1] 100\n",
      "[1] 101\n",
      "[1] 102\n",
      "[1] 103\n",
      "[1] 104\n",
      "[1] 105\n",
      "[1] 106\n",
      "[1] 107\n",
      "[1] 108\n",
      "[1] 109\n",
      "[1] 110\n",
      "[1] 111\n",
      "[1] 112\n",
      "[1] 113\n",
      "[1] 114\n",
      "[1] 115\n",
      "[1] 116\n",
      "[1] 117\n",
      "[1] 118\n",
      "[1] 119\n",
      "[1] 120\n",
      "[1] 121\n",
      "[1] 122\n",
      "[1] 123\n",
      "[1] 124\n",
      "[1] 125\n",
      "[1] 126\n",
      "[1] 127\n",
      "[1] 128\n",
      "[1] 129\n",
      "[1] 130\n",
      "[1] 131\n",
      "[1] 132\n",
      "[1] 133\n",
      "[1] 134\n",
      "[1] 135\n",
      "[1] 136\n",
      "[1] 137\n",
      "[1] 138\n",
      "[1] 139\n",
      "[1] 140\n",
      "[1] 141\n",
      "[1] 142\n",
      "[1] 143\n",
      "[1] 144\n",
      "[1] 145\n",
      "[1] 146\n",
      "[1] 147\n",
      "[1] 148\n",
      "[1] 149\n",
      "[1] 150\n",
      "[1] 151\n",
      "[1] 152\n",
      "[1] 153\n",
      "[1] 154\n",
      "[1] 155\n",
      "[1] 156\n",
      "[1] 157\n",
      "[1] 158\n",
      "[1] 159\n",
      "[1] 160\n",
      "[1] 161\n",
      "[1] 162\n",
      "[1] 163\n",
      "[1] 164\n",
      "[1] 165\n",
      "[1] 166\n",
      "[1] 167\n",
      "[1] 168\n",
      "[1] 169\n",
      "[1] 170\n",
      "[1] 171\n",
      "[1] 172\n",
      "[1] 173\n",
      "[1] 174\n",
      "[1] 175\n",
      "[1] 176\n",
      "[1] 177\n",
      "[1] 178\n",
      "[1] 179\n",
      "[1] 180\n",
      "[1] 181\n",
      "[1] 182\n",
      "[1] 183\n",
      "[1] 184\n",
      "[1] 185\n",
      "[1] 186\n",
      "[1] 187\n",
      "[1] 188\n",
      "[1] 189\n",
      "[1] 190\n",
      "[1] 191\n",
      "[1] 192\n",
      "[1] 193\n",
      "[1] 194\n",
      "[1] 195\n",
      "[1] 196\n",
      "[1] 197\n",
      "[1] 198\n",
      "[1] 199\n",
      "[1] 200\n",
      "[1] 201\n",
      "[1] 202\n",
      "[1] 203\n",
      "[1] 204\n",
      "[1] 205\n",
      "[1] 206\n",
      "[1] 207\n",
      "[1] 208\n",
      "[1] 209\n",
      "[1] 210\n",
      "[1] 211\n",
      "[1] 212\n",
      "[1] 213\n",
      "[1] 214\n",
      "[1] 215\n",
      "[1] 216\n",
      "[1] 217\n",
      "[1] 218\n",
      "[1] 219\n",
      "[1] 220\n",
      "[1] 221\n",
      "[1] 222\n",
      "[1] 223\n",
      "[1] 224\n",
      "[1] 225\n",
      "[1] 226\n",
      "[1] 227\n",
      "[1] 228\n",
      "[1] 229\n",
      "[1] 230\n",
      "[1] 231\n",
      "[1] 232\n",
      "[1] 233\n",
      "[1] 234\n",
      "[1] 235\n",
      "[1] 236\n",
      "[1] 237\n",
      "[1] 238\n",
      "[1] 239\n",
      "[1] 240\n",
      "[1] 241\n",
      "[1] 242\n",
      "[1] 243\n",
      "[1] 244\n",
      "[1] 245\n",
      "[1] 246\n",
      "[1] 247\n",
      "[1] 248\n",
      "[1] 249\n",
      "[1] 250\n",
      "[1] 251\n",
      "[1] 252\n",
      "[1] 253\n",
      "[1] 254\n",
      "[1] 255\n",
      "[1] 256\n",
      "[1] 257\n",
      "[1] 258\n",
      "[1] 259\n",
      "[1] 260\n",
      "[1] 261\n",
      "[1] 262\n",
      "[1] 263\n",
      "[1] 264\n",
      "[1] 265\n",
      "[1] 266\n",
      "[1] 267\n",
      "[1] 268\n",
      "[1] 269\n",
      "[1] 270\n",
      "[1] 271\n",
      "[1] 272\n",
      "[1] 273\n",
      "[1] 274\n",
      "[1] 275\n",
      "[1] 276\n",
      "[1] 277\n",
      "[1] 278\n",
      "[1] 279\n",
      "[1] 280\n",
      "[1] 281\n",
      "[1] 282\n",
      "[1] 283\n",
      "[1] 284\n",
      "[1] 285\n",
      "[1] 286\n",
      "[1] 287\n",
      "[1] 288\n",
      "[1] 289\n",
      "[1] 290\n",
      "[1] 291\n",
      "[1] 292\n",
      "[1] 293\n",
      "[1] 294\n",
      "[1] 295\n",
      "[1] 296\n",
      "[1] 297\n",
      "[1] 298\n",
      "[1] 299\n",
      "[1] 300\n",
      "[1] 301\n",
      "[1] 302\n",
      "[1] 303\n",
      "[1] 304\n",
      "[1] 305\n",
      "[1] 306\n",
      "[1] 307\n",
      "[1] 308\n",
      "[1] 309\n",
      "[1] 310\n",
      "[1] 311\n",
      "[1] 312\n",
      "[1] 313\n",
      "[1] 314\n",
      "[1] 315\n",
      "[1] 316\n",
      "[1] 317\n",
      "[1] 318\n",
      "[1] 319\n",
      "[1] 320\n",
      "[1] 321\n",
      "[1] 322\n",
      "[1] 323\n",
      "[1] 324\n",
      "[1] 325\n",
      "[1] 326\n",
      "[1] 327\n",
      "[1] 328\n",
      "[1] 329\n",
      "[1] 330\n",
      "[1] 331\n",
      "[1] 332\n",
      "[1] 333\n",
      "[1] 334\n",
      "[1] 335\n",
      "[1] 336\n",
      "[1] 337\n",
      "[1] 338\n",
      "[1] 339\n",
      "[1] 340\n",
      "[1] 341\n",
      "[1] 342\n",
      "[1] 343\n",
      "[1] 344\n",
      "[1] 345\n",
      "[1] 346\n",
      "[1] 347\n",
      "[1] 348\n",
      "[1] 349\n",
      "[1] 350\n",
      "[1] 351\n",
      "[1] 352\n",
      "[1] 353\n",
      "[1] 354\n",
      "[1] 355\n",
      "[1] 356\n",
      "[1] 357\n",
      "[1] 358\n",
      "[1] 359\n",
      "[1] 360\n",
      "[1] 361\n",
      "[1] 362\n",
      "[1] 363\n",
      "[1] 364\n",
      "[1] 365\n",
      "[1] 366\n",
      "[1] 367\n",
      "[1] 368\n",
      "[1] 369\n",
      "[1] 370\n",
      "[1] 371\n",
      "[1] 372\n",
      "[1] 373\n",
      "[1] 374\n",
      "[1] 375\n",
      "[1] 376\n",
      "[1] 377\n",
      "[1] 378\n",
      "[1] 379\n",
      "[1] 380\n",
      "[1] 381\n",
      "[1] 382\n",
      "[1] 383\n",
      "[1] 384\n",
      "[1] 385\n",
      "[1] 386\n",
      "[1] 387\n",
      "[1] 388\n",
      "[1] 389\n",
      "[1] 390\n",
      "[1] 391\n",
      "[1] 392\n",
      "[1] 393\n",
      "[1] 394\n",
      "[1] 395\n",
      "[1] 396\n",
      "[1] 397\n",
      "[1] 398\n",
      "[1] 399\n",
      "[1] 400\n",
      "[1] 401\n",
      "[1] 402\n",
      "[1] 403\n",
      "[1] 404\n",
      "[1] 405\n",
      "[1] 406\n",
      "[1] 407\n",
      "[1] 408\n",
      "[1] 409\n",
      "[1] 410\n",
      "[1] 411\n",
      "[1] 412\n",
      "[1] 413\n",
      "[1] 414\n",
      "[1] 415\n",
      "[1] 416\n",
      "[1] 417\n",
      "[1] 418\n",
      "[1] 419\n",
      "[1] 420\n",
      "[1] 421\n",
      "[1] 422\n",
      "[1] 423\n",
      "[1] 424\n",
      "[1] 425\n",
      "[1] 426\n",
      "[1] 427\n",
      "[1] 428\n",
      "[1] 429\n",
      "[1] 430\n",
      "[1] 431\n",
      "[1] 432\n",
      "[1] 433\n",
      "[1] 434\n",
      "[1] 435\n",
      "[1] 436\n",
      "[1] 437\n",
      "[1] 438\n",
      "[1] 439\n",
      "[1] 440\n",
      "[1] 441\n",
      "[1] 442\n",
      "[1] 443\n",
      "[1] 444\n",
      "[1] 445\n",
      "[1] 446\n",
      "[1] 447\n",
      "[1] 448\n",
      "[1] 449\n",
      "[1] 450\n",
      "[1] 451\n",
      "[1] 452\n",
      "[1] 453\n",
      "[1] 454\n",
      "[1] 455\n",
      "[1] 456\n",
      "[1] 457\n",
      "[1] 458\n",
      "[1] 459\n",
      "[1] 460\n",
      "[1] 461\n",
      "[1] 462\n",
      "[1] 463\n",
      "[1] 464\n",
      "[1] 465\n",
      "[1] 466\n",
      "[1] 467\n",
      "[1] 468\n",
      "[1] 469\n",
      "[1] 470\n",
      "[1] 471\n",
      "[1] 472\n",
      "[1] 473\n",
      "[1] 474\n",
      "[1] 475\n",
      "[1] 476\n",
      "[1] 477\n",
      "[1] 478\n",
      "[1] 479\n",
      "[1] 480\n",
      "[1] 481\n",
      "[1] 482\n",
      "[1] 483\n",
      "[1] 484\n",
      "[1] 485\n",
      "[1] 486\n",
      "[1] 487\n",
      "[1] 488\n",
      "[1] 489\n",
      "[1] 490\n",
      "[1] 491\n",
      "[1] 492\n",
      "[1] 493\n",
      "[1] 494\n",
      "[1] 495\n",
      "[1] 496\n",
      "[1] 497\n",
      "[1] 498\n",
      "[1] 499\n",
      "[1] 500\n",
      "[1] 501\n",
      "[1] 502\n",
      "[1] 503\n",
      "[1] 504\n",
      "[1] 505\n",
      "[1] 506\n",
      "[1] 507\n",
      "[1] 508\n",
      "[1] 509\n",
      "[1] 510\n",
      "[1] 511\n",
      "[1] 512\n",
      "[1] 513\n",
      "[1] 514\n",
      "[1] 515\n",
      "[1] 516\n",
      "[1] 517\n",
      "[1] 518\n",
      "[1] 519\n",
      "[1] 520\n",
      "[1] 521\n",
      "[1] 522\n",
      "[1] 523\n",
      "[1] 524\n",
      "[1] 525\n",
      "[1] 526\n",
      "[1] 527\n",
      "[1] 528\n",
      "[1] 529\n",
      "[1] 530\n",
      "[1] 531\n",
      "[1] 532\n",
      "[1] 533\n",
      "[1] 534\n",
      "[1] 535\n",
      "[1] 536\n",
      "[1] 537\n",
      "[1] 538\n",
      "[1] 539\n",
      "[1] 540\n",
      "[1] 541\n",
      "[1] 542\n",
      "[1] 543\n",
      "[1] 544\n",
      "[1] 545\n",
      "[1] 546\n",
      "[1] 547\n",
      "[1] 548\n",
      "[1] 549\n",
      "[1] 550\n",
      "[1] 551\n",
      "[1] 552\n",
      "[1] 553\n",
      "[1] 554\n",
      "[1] 555\n",
      "[1] 556\n",
      "[1] 557\n",
      "[1] 558\n",
      "[1] 559\n",
      "[1] 560\n",
      "[1] 561\n",
      "[1] 562\n",
      "[1] 563\n",
      "[1] 564\n",
      "[1] 565\n",
      "[1] 566\n",
      "[1] 567\n",
      "[1] 568\n",
      "[1] 569\n",
      "[1] 570\n",
      "[1] 571\n",
      "[1] 572\n",
      "[1] 573\n",
      "[1] 574\n",
      "[1] 575\n",
      "[1] 576\n",
      "[1] 577\n",
      "[1] 578\n",
      "[1] 579\n",
      "[1] 580\n",
      "[1] 581\n",
      "[1] 582\n",
      "[1] 583\n",
      "[1] 584\n",
      "[1] 585\n",
      "[1] 586\n",
      "[1] 587\n",
      "[1] 588\n",
      "[1] 589\n",
      "[1] 590\n",
      "[1] 591\n",
      "[1] 592\n",
      "[1] 593\n",
      "[1] 594\n",
      "[1] 595\n",
      "[1] 596\n",
      "[1] 597\n",
      "[1] 598\n",
      "[1] 599\n",
      "[1] 600\n",
      "[1] 601\n",
      "[1] 602\n",
      "[1] 603\n",
      "[1] 604\n",
      "[1] 605\n",
      "[1] 606\n",
      "[1] 607\n",
      "[1] 608\n",
      "[1] 609\n",
      "[1] 610\n",
      "[1] 611\n",
      "[1] 612\n",
      "[1] 613\n",
      "[1] 614\n",
      "[1] 615\n",
      "[1] 616\n",
      "[1] 617\n",
      "[1] 618\n",
      "[1] 619\n",
      "[1] 620\n",
      "[1] 621\n",
      "[1] 622\n",
      "[1] 623\n",
      "[1] 624\n",
      "[1] 625\n",
      "[1] 626\n",
      "[1] 627\n",
      "[1] 628\n",
      "[1] 629\n",
      "[1] 630\n",
      "[1] 631\n",
      "[1] 632\n",
      "[1] 633\n",
      "[1] 634\n",
      "[1] 635\n",
      "[1] 636\n",
      "[1] 637\n",
      "[1] 638\n",
      "[1] 639\n",
      "[1] 640\n",
      "[1] 641\n",
      "[1] 642\n",
      "[1] 643\n",
      "[1] 644\n",
      "[1] 645\n",
      "[1] 646\n",
      "[1] 647\n",
      "[1] 648\n",
      "[1] 649\n",
      "[1] 650\n",
      "[1] 651\n",
      "[1] 652\n",
      "[1] 653\n",
      "[1] 654\n",
      "[1] 655\n",
      "[1] 656\n",
      "[1] 657\n",
      "[1] 658\n",
      "[1] 659\n",
      "[1] 660\n",
      "[1] 661\n",
      "[1] 662\n",
      "[1] 663\n",
      "[1] 664\n",
      "[1] 665\n",
      "[1] 666\n",
      "[1] 667\n",
      "[1] 668\n",
      "[1] 669\n",
      "[1] 670\n",
      "[1] 671\n",
      "[1] 672\n",
      "[1] 673\n",
      "[1] 674\n",
      "[1] 675\n",
      "[1] 676\n",
      "[1] 677\n",
      "[1] 678\n",
      "[1] 679\n",
      "[1] 680\n",
      "[1] 681\n",
      "[1] 682\n",
      "[1] 683\n",
      "[1] 684\n",
      "[1] 685\n",
      "[1] 686\n",
      "[1] 687\n",
      "[1] 688\n",
      "[1] 689\n",
      "[1] 690\n",
      "[1] 691\n",
      "[1] 692\n",
      "[1] 693\n",
      "[1] 694\n",
      "[1] 695\n",
      "[1] 696\n",
      "[1] 697\n",
      "[1] 698\n",
      "[1] 699\n",
      "[1] 700\n",
      "[1] 701\n",
      "[1] 702\n",
      "[1] 703\n",
      "[1] 704\n",
      "[1] 705\n",
      "[1] 706\n",
      "[1] 707\n",
      "[1] 708\n",
      "[1] 709\n",
      "[1] 710\n",
      "[1] 711\n",
      "[1] 712\n",
      "[1] 713\n",
      "[1] 714\n",
      "[1] 715\n",
      "[1] 716\n",
      "[1] 717\n",
      "[1] 718\n",
      "[1] 719\n",
      "[1] 720\n",
      "[1] 721\n",
      "[1] 722\n",
      "[1] 723\n",
      "[1] 724\n",
      "[1] 725\n",
      "[1] 726\n",
      "[1] 727\n",
      "[1] 728\n",
      "[1] 729\n",
      "[1] 730\n",
      "[1] 731\n",
      "[1] 732\n",
      "[1] 733\n",
      "[1] 734\n",
      "[1] 735\n",
      "[1] 736\n",
      "[1] 737\n",
      "[1] 738\n",
      "[1] 739\n",
      "[1] 740\n",
      "[1] 741\n",
      "[1] 742\n",
      "[1] 743\n",
      "[1] 744\n",
      "[1] 745\n",
      "[1] 746\n",
      "[1] 747\n",
      "[1] 748\n",
      "[1] 749\n",
      "[1] 750\n",
      "[1] 751\n",
      "[1] 752\n",
      "[1] 753\n",
      "[1] 754\n",
      "[1] 755\n",
      "[1] 756\n",
      "[1] 757\n",
      "[1] 758\n",
      "[1] 759\n",
      "[1] 760\n",
      "[1] 761\n",
      "[1] 762\n",
      "[1] 763\n",
      "[1] 764\n",
      "[1] 765\n",
      "[1] 766\n",
      "[1] 767\n",
      "[1] 768\n",
      "[1] 769\n",
      "[1] 770\n",
      "[1] 771\n",
      "[1] 772\n",
      "[1] 773\n",
      "[1] 774\n",
      "[1] 775\n",
      "[1] 776\n",
      "[1] 777\n",
      "[1] 778\n",
      "[1] 779\n",
      "[1] 780\n",
      "[1] 781\n",
      "[1] 782\n",
      "[1] 783\n",
      "[1] 784\n",
      "[1] 785\n",
      "[1] 786\n",
      "[1] 787\n",
      "[1] 788\n",
      "[1] 789\n",
      "[1] 790\n",
      "[1] 791\n",
      "[1] 792\n",
      "[1] 793\n",
      "[1] 794\n",
      "[1] 795\n",
      "[1] 796\n",
      "[1] 797\n",
      "[1] 798\n",
      "[1] 799\n",
      "[1] 800\n",
      "[1] 801\n",
      "[1] 802\n",
      "[1] 803\n",
      "[1] 804\n",
      "[1] 805\n",
      "[1] 806\n",
      "[1] 807\n",
      "[1] 808\n",
      "[1] 809\n",
      "[1] 810\n",
      "[1] 811\n",
      "[1] 812\n",
      "[1] 813\n",
      "[1] 814\n",
      "[1] 815\n",
      "[1] 816\n",
      "[1] 817\n",
      "[1] 818\n",
      "[1] 819\n",
      "[1] 820\n",
      "[1] 821\n",
      "[1] 822\n",
      "[1] 823\n",
      "[1] 824\n",
      "[1] 825\n",
      "[1] 826\n",
      "[1] 827\n",
      "[1] 828\n",
      "[1] 829\n",
      "[1] 830\n",
      "[1] 831\n",
      "[1] 832\n",
      "[1] 833\n",
      "[1] 834\n",
      "[1] 835\n",
      "[1] 836\n",
      "[1] 837\n",
      "[1] 838\n",
      "[1] 839\n",
      "[1] 840\n",
      "[1] 841\n",
      "[1] 842\n",
      "[1] 843\n",
      "[1] 844\n",
      "[1] 845\n",
      "[1] 846\n",
      "[1] 847\n",
      "[1] 848\n",
      "[1] 849\n",
      "[1] 850\n",
      "[1] 851\n",
      "[1] 852\n",
      "[1] 853\n",
      "[1] 854\n",
      "[1] 855\n",
      "[1] 856\n",
      "[1] 857\n",
      "[1] 858\n",
      "[1] 859\n",
      "[1] 860\n",
      "[1] 861\n",
      "[1] 862\n",
      "[1] 863\n",
      "[1] 864\n",
      "[1] 865\n",
      "[1] 866\n",
      "[1] 867\n",
      "[1] 868\n",
      "[1] 869\n",
      "[1] 870\n",
      "[1] 871\n",
      "[1] 872\n",
      "[1] 873\n",
      "[1] 874\n",
      "[1] 875\n",
      "[1] 876\n",
      "[1] 877\n",
      "[1] 878\n",
      "[1] 879\n",
      "[1] 880\n",
      "[1] 881\n",
      "[1] 882\n",
      "[1] 883\n",
      "[1] 884\n",
      "[1] 885\n",
      "[1] 886\n",
      "[1] 887\n",
      "[1] 888\n",
      "[1] 889\n",
      "[1] 890\n",
      "[1] 891\n",
      "[1] 892\n",
      "[1] 893\n",
      "[1] 894\n",
      "[1] 895\n",
      "[1] 896\n",
      "[1] 897\n",
      "[1] 898\n",
      "[1] 899\n",
      "[1] 900\n",
      "[1] 901\n",
      "[1] 902\n",
      "[1] 903\n",
      "[1] 904\n",
      "[1] 905\n",
      "[1] 906\n",
      "[1] 907\n",
      "[1] 908\n",
      "[1] 909\n",
      "[1] 910\n",
      "[1] 911\n",
      "[1] 912\n",
      "[1] 913\n",
      "[1] 914\n",
      "[1] 915\n",
      "[1] 916\n",
      "[1] 917\n",
      "[1] 918\n",
      "[1] 919\n",
      "[1] 920\n",
      "[1] 921\n",
      "[1] 922\n",
      "[1] 923\n",
      "[1] 924\n",
      "[1] 925\n",
      "[1] 926\n",
      "[1] 927\n",
      "[1] 928\n",
      "[1] 929\n",
      "[1] 930\n",
      "[1] 931\n",
      "[1] 932\n",
      "[1] 933\n",
      "[1] 934\n",
      "[1] 935\n",
      "[1] 936\n",
      "[1] 937\n",
      "[1] 938\n",
      "[1] 939\n",
      "[1] 940\n",
      "[1] 941\n",
      "[1] 942\n",
      "[1] 943\n",
      "[1] 944\n",
      "[1] 945\n",
      "[1] 946\n",
      "[1] 947\n",
      "[1] 948\n",
      "[1] 949\n",
      "[1] 950\n",
      "[1] 951\n",
      "[1] 952\n",
      "[1] 953\n",
      "[1] 954\n",
      "[1] 955\n",
      "[1] 956\n",
      "[1] 957\n",
      "[1] 958\n",
      "[1] 959\n",
      "[1] 960\n",
      "[1] 961\n",
      "[1] 962\n",
      "[1] 963\n",
      "[1] 964\n",
      "[1] 965\n",
      "[1] 966\n",
      "[1] 967\n",
      "[1] 968\n",
      "[1] 969\n",
      "[1] 970\n",
      "[1] 971\n",
      "[1] 972\n",
      "[1] 973\n",
      "[1] 974\n",
      "[1] 975\n",
      "[1] 976\n",
      "[1] 977\n",
      "[1] 978\n",
      "[1] 979\n",
      "[1] 980\n",
      "[1] 981\n",
      "[1] 982\n",
      "[1] 983\n",
      "[1] 984\n",
      "[1] 985\n",
      "[1] 986\n",
      "[1] 987\n",
      "[1] 988\n",
      "[1] 989\n",
      "[1] 990\n",
      "[1] 991\n",
      "[1] 992\n",
      "[1] 993\n",
      "[1] 994\n",
      "[1] 995\n",
      "[1] 996\n",
      "[1] 997\n",
      "[1] 998\n",
      "[1] 999\n",
      "[1] 1000\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A matrix: 12 × 2 of type chr</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>A</th><th scope=col>B</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>ANKRD22</td><td>CA5B  </td></tr>\n",
       "\t<tr><td>CEACAM1</td><td>FCMR  </td></tr>\n",
       "\t<tr><td>FCGR1A </td><td>GZMK  </td></tr>\n",
       "\t<tr><td>FCGR1BP</td><td>SIRPG </td></tr>\n",
       "\t<tr><td>GBP1   </td><td>SLAMF6</td></tr>\n",
       "\t<tr><td>GBP4   </td><td>SS18L2</td></tr>\n",
       "\t<tr><td>GBP5   </td><td>CA5B  </td></tr>\n",
       "\t<tr><td>GBP6   </td><td>FCMR  </td></tr>\n",
       "\t<tr><td>LHFPL2 </td><td>GZMK  </td></tr>\n",
       "\t<tr><td>PSMB9  </td><td>SIRPG </td></tr>\n",
       "\t<tr><td>PSTPIP2</td><td>SLAMF6</td></tr>\n",
       "\t<tr><td>VAMP5  </td><td>SS18L2</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A matrix: 12 × 2 of type chr\n",
       "\\begin{tabular}{ll}\n",
       " A & B\\\\\n",
       "\\hline\n",
       "\t ANKRD22 & CA5B  \\\\\n",
       "\t CEACAM1 & FCMR  \\\\\n",
       "\t FCGR1A  & GZMK  \\\\\n",
       "\t FCGR1BP & SIRPG \\\\\n",
       "\t GBP1    & SLAMF6\\\\\n",
       "\t GBP4    & SS18L2\\\\\n",
       "\t GBP5    & CA5B  \\\\\n",
       "\t GBP6    & FCMR  \\\\\n",
       "\t LHFPL2  & GZMK  \\\\\n",
       "\t PSMB9   & SIRPG \\\\\n",
       "\t PSTPIP2 & SLAMF6\\\\\n",
       "\t VAMP5   & SS18L2\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A matrix: 12 × 2 of type chr\n",
       "\n",
       "| A | B |\n",
       "|---|---|\n",
       "| ANKRD22 | CA5B   |\n",
       "| CEACAM1 | FCMR   |\n",
       "| FCGR1A  | GZMK   |\n",
       "| FCGR1BP | SIRPG  |\n",
       "| GBP1    | SLAMF6 |\n",
       "| GBP4    | SS18L2 |\n",
       "| GBP5    | CA5B   |\n",
       "| GBP6    | FCMR   |\n",
       "| LHFPL2  | GZMK   |\n",
       "| PSMB9   | SIRPG  |\n",
       "| PSTPIP2 | SLAMF6 |\n",
       "| VAMP5   | SS18L2 |\n",
       "\n"
      ],
      "text/plain": [
       "      A       B     \n",
       " [1,] ANKRD22 CA5B  \n",
       " [2,] CEACAM1 FCMR  \n",
       " [3,] FCGR1A  GZMK  \n",
       " [4,] FCGR1BP SIRPG \n",
       " [5,] GBP1    SLAMF6\n",
       " [6,] GBP4    SS18L2\n",
       " [7,] GBP5    CA5B  \n",
       " [8,] GBP6    FCMR  \n",
       " [9,] LHFPL2  GZMK  \n",
       "[10,] PSMB9   SIRPG \n",
       "[11,] PSTPIP2 SLAMF6\n",
       "[12,] VAMP5   SS18L2"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Call function\n",
    "train_data = read.csv(\"./data/03_Final_data/01_train_data.csv\", check.names = FALSE, row.names = 1)\n",
    "train_labels = as.factor(train_data[, ncol(train_data)])\n",
    "train_data = train_data[, -ncol(train_data)]\n",
    "train_data = t(train_data)\n",
    "percentage = 0.9\n",
    "features_num = 20\n",
    "iterations_num = 1000\n",
    "freq_threshold = 0.8\n",
    "output_file = \"./result/04_TSPG/01_selected_genes.csv\"\n",
    "extract_genes(train_data, train_labels, percentage, features_num, iterations_num, freq_threshold, output_file)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0bff609a-0c2a-472d-b8c0-1bd152539122",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "#### 3.2.2 Test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "fe6e946f-8601-48ac-85a2-29a0e7691c55",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-02T08:05:01.273949Z",
     "iopub.status.busy": "2024-12-02T08:05:01.271816Z",
     "iopub.status.idle": "2024-12-02T08:05:01.295311Z",
     "shell.execute_reply": "2024-12-02T08:05:01.291890Z"
    }
   },
   "outputs": [],
   "source": [
    "# Define function\n",
    "rank_based_classifier = function(data, gene_matrix, labels) {\n",
    "  # Extract relevant gene matrix\n",
    "  all_genes = c(gene_matrix[, 1], gene_matrix[, 2])\n",
    "  data_subset = data[all_genes, , drop = FALSE]\n",
    "  \n",
    "  # Initialize prediction result vector\n",
    "  predictions = numeric(ncol(data))\n",
    "  \n",
    "  # Iterate through each sample\n",
    "  for (i in 1:ncol(data)) {\n",
    "    print(i)\n",
    "    flush.console()\n",
    "      \n",
    "    # Extract gene data for each sample\n",
    "    sample_data = data_subset[, i]\n",
    "    \n",
    "    # Apply rank transformation\n",
    "    ranked_data = rank(sample_data)\n",
    "    \n",
    "    # Get average rank value\n",
    "    A_genes = gene_matrix[, 1]\n",
    "    B_genes = gene_matrix[, 2]\n",
    "    A_ranks = ranked_data[A_genes]\n",
    "    B_ranks = ranked_data[B_genes]\n",
    "    A_mean_rank = mean(A_ranks, na.rm = TRUE)\n",
    "    B_mean_rank = mean(B_ranks, na.rm = TRUE)\n",
    "    \n",
    "    # Make prediction\n",
    "    if (A_mean_rank < B_mean_rank) {\n",
    "      predictions[i] = 0\n",
    "    } else {\n",
    "      predictions[i] = 1\n",
    "    }\n",
    "  }\n",
    "\n",
    "  accuracy <- sum(predictions == labels) / length(labels)\n",
    "  return(accuracy)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "efb039fd-7667-42bd-81f4-f975337dc6a3",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-02T08:05:01.300401Z",
     "iopub.status.busy": "2024-12-02T08:05:01.298111Z",
     "iopub.status.idle": "2024-12-02T08:05:24.859065Z",
     "shell.execute_reply": "2024-12-02T08:05:24.855026Z"
    },
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 1\n",
      "[1] 2\n",
      "[1] 3\n",
      "[1] 4\n",
      "[1] 5\n",
      "[1] 6\n",
      "[1] 7\n",
      "[1] 8\n",
      "[1] 9\n",
      "[1] 10\n",
      "[1] 11\n",
      "[1] 12\n",
      "[1] 13\n",
      "[1] 14\n",
      "[1] 15\n",
      "[1] 16\n",
      "[1] 17\n",
      "[1] 18\n",
      "[1] 19\n",
      "[1] 20\n",
      "[1] 21\n",
      "[1] 22\n",
      "[1] 23\n",
      "[1] 24\n",
      "[1] 25\n",
      "[1] 26\n",
      "[1] 27\n",
      "[1] 28\n",
      "[1] 29\n",
      "[1] 30\n",
      "[1] 31\n",
      "[1] 32\n",
      "[1] 33\n",
      "[1] 34\n",
      "[1] 35\n",
      "[1] 36\n",
      "[1] 37\n",
      "[1] 38\n",
      "[1] 39\n",
      "[1] 40\n",
      "[1] 41\n",
      "[1] 42\n",
      "[1] 43\n",
      "[1] 44\n",
      "[1] 45\n",
      "[1] 46\n",
      "[1] 47\n",
      "[1] 48\n",
      "[1] 49\n",
      "[1] 50\n",
      "[1] 51\n",
      "[1] 52\n",
      "[1] 53\n",
      "[1] 54\n",
      "[1] 55\n",
      "[1] 56\n",
      "[1] 57\n",
      "[1] 58\n",
      "[1] 59\n",
      "[1] 60\n",
      "[1] 61\n",
      "[1] 62\n",
      "[1] 63\n",
      "[1] 64\n",
      "[1] 65\n",
      "[1] 66\n",
      "[1] 67\n",
      "[1] 68\n",
      "[1] 69\n",
      "[1] 70\n",
      "[1] 71\n",
      "[1] 72\n",
      "[1] 73\n",
      "[1] 74\n",
      "[1] 75\n",
      "[1] 76\n",
      "[1] 77\n",
      "[1] 78\n",
      "[1] 79\n",
      "[1] 80\n",
      "[1] 81\n",
      "[1] 82\n",
      "[1] 83\n",
      "[1] 84\n",
      "[1] 85\n",
      "[1] 86\n",
      "[1] 87\n",
      "[1] 88\n",
      "[1] 89\n",
      "[1] 90\n",
      "[1] 91\n",
      "[1] 92\n",
      "[1] 93\n",
      "[1] 94\n",
      "[1] 95\n",
      "[1] 96\n",
      "[1] 97\n",
      "[1] 98\n",
      "[1] 99\n",
      "[1] 100\n",
      "[1] 101\n",
      "[1] 102\n",
      "[1] 103\n",
      "[1] 104\n",
      "[1] 105\n",
      "[1] 106\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "0.924528301886792"
      ],
      "text/latex": [
       "0.924528301886792"
      ],
      "text/markdown": [
       "0.924528301886792"
      ],
      "text/plain": [
       "[1] 0.9245283"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Call function\n",
    "data = read.csv(\"./data/03_Final_data/03_valid_data.csv\", check.names = FALSE, row.names = 1)\n",
    "labels = as.factor(data[, ncol(data)])\n",
    "data = data[, -ncol(data)]\n",
    "data = t(data)\n",
    "gene_matrix = read.csv(\"./result/04_TSPG/01_selected_genes.csv\", check.names = FALSE, row.names = 1)\n",
    "rank_based_classifier(data, gene_matrix, labels)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9584cff6-e436-4cf6-b66f-0d15503ca4e3",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-02T06:58:52.536543Z",
     "iopub.status.busy": "2024-12-02T06:58:52.533408Z",
     "iopub.status.idle": "2024-12-02T06:58:52.557751Z",
     "shell.execute_reply": "2024-12-02T06:58:52.554480Z"
    }
   },
   "source": [
    "**train:0.913 ; test:0.886 ; valid:0.925**"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "40a482ff-a9e9-4207-b11f-9c1c88029e3a",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "# 4、Plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "28357c65-3129-4a08-b8be-919b537eb36a",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-03T11:51:48.847694Z",
     "iopub.status.busy": "2024-12-03T11:51:48.846392Z",
     "iopub.status.idle": "2024-12-03T11:51:48.859328Z",
     "shell.execute_reply": "2024-12-03T11:51:48.857796Z"
    }
   },
   "outputs": [],
   "source": [
    "# Define function\n",
    "create_polar_bar_chart = function(input_file, y_limit, y_breaks, colors, output_file) {\n",
    "  # Read data\n",
    "  data = read.csv(input_file, sep = ',', header = TRUE, stringsAsFactors = FALSE)\n",
    "    \n",
    "  # Add order column\n",
    "  data = data %>% mutate(Order = 1:n()) # Add a column for specifying order\n",
    "  data$Order = factor(data$Order) # Convert to factor type for color assign\n",
    "  \n",
    "  # Plot polar bar chart\n",
    "  p = data %>% ggplot(aes(x = Order, y = ACC, fill = Order)) + \n",
    "    geom_bar(stat = 'identity') + # Plot bars with height using raw values, no additional aggregation\n",
    "    coord_polar(theta = \"y\") + # Apply polar transformation\n",
    "    theme_bw() + # Remove background\n",
    "    theme_minimal() + # Remove borders\n",
    "    xlab(\"\") + ylab(\"\") + # Remove x and y axis labels\n",
    "    # Remove legend, y-axis ticks, and minor grid lines\n",
    "    theme(legend.position = \"none\", axis.text.y = element_blank(), panel.grid.minor = element_blank()) + \n",
    "    # Add labels: content, position, font, size, alignment\n",
    "    geom_text(aes(label = Method, x = Order, y = 0, fontface = \"italic\"), size = 3, hjust = 1) +\n",
    "    # Set range ticks\n",
    "    scale_y_continuous(limits = c(0, y_limit), breaks = seq(0, y_limit, by = y_breaks)) + \n",
    "    scale_fill_manual(values = colors) # Set fill color\n",
    "  \n",
    "  # Save as PDF\n",
    "  ggsave(output_file, plot = p, height = 6, width = 6, dpi = 300)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "80c30243-9671-45cb-9a44-7d44c3b19ebb",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-03T11:53:55.346006Z",
     "iopub.status.busy": "2024-12-03T11:53:55.344812Z",
     "iopub.status.idle": "2024-12-03T11:53:55.477683Z",
     "shell.execute_reply": "2024-12-03T11:53:55.476465Z"
    }
   },
   "outputs": [],
   "source": [
    "# Run\n",
    "input_file = './result/09_plot/valid_ACC.csv'\n",
    "y_limit = 1\n",
    "y_breaks = 0.1\n",
    "colors = c(\"#ffef1e\", \"#fdfa0a\", \"#deff02\", \"#c9ff00\", \"#b7ff02\", \"#82e0aa\", \"#7dcea0\", \"#73c6b6\", \"#76d7c4\")\n",
    "output_file = './result/09_plot/03_valid_ACC.pdf'\n",
    "create_polar_bar_chart(input_file, y_limit, y_breaks, colors, output_file)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R (4.2.3)",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.2.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
